File: orchalang/src/main/java/orcha/lang/compiler/OrchaMetadata.java
package orcha.lang.compiler;

import orcha.lang.compiler.syntax.Instruction;
import orcha.lang.compiler.syntax.TitleInstruction;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class OrchaMetadata {

    List<Instruction> metadata = new ArrayList<Instruction>();

    String title;
    String description;
    String domain;
    String author;
    String version;

    public void add(Instruction instruction) {
        metadata.add(instruction);
    }

    public List<Instruction> getMetadata() {
        return metadata;
    }

    public String getTitle() {
        TitleInstruction titleInstruction = (TitleInstruction) metadata.stream().filter(instruction -> instruction instanceof TitleInstruction).findAny().orElse(null);
        if(titleInstruction != null){
            return titleInstruction.getTitle();
        }
        return null;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getDomain() {
        return domain;
    }

    public void setDomain(String domain) {
        this.domain = domain;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    @Override
    public String toString() {
        return "OrchaMetadata{" +
                "title='" + title + '\'' +
                ", description='" + description + '\'' +
                ", domain='" + domain + '\'' +
                ", author='" + author + '\'' +
                ", version='" + version + '\'' +
                '}';
    }

}

File: orchalang/src/main/kotlin/orcha/lang/compiler/OrchaMetadata.kt
package orcha.lang.compiler

import orcha.lang.compiler.syntax.*
import java.util.*

data class OrchaMetadata(var metadata: MutableList<Instruction?> = ArrayList()) {

    var authors: MutableList<String>? = null
        get() {
            val authorInstruction = metadata.stream().filter { instruction: Instruction? -> instruction is AuthorsInstruction }.findAny().orElse(null) as AuthorsInstruction?
            return authorInstruction?.authors
        }

    var description: String? = null
        get() {
            val descriptionInstruction = metadata.stream().filter { instruction: Instruction? -> instruction is DescriptionInstruction }.findAny().orElse(null) as DescriptionInstruction?
            return descriptionInstruction?.description
        }

    var domain: String? = null
        get() {
            val domainInstruction = metadata.stream().filter { instruction: Instruction? -> instruction is DomainInstruction }.findAny().orElse(null) as DomainInstruction?
            return domainInstruction?.domain
        }

    var domainAsCapitalizedConcatainedString: String? = null
        get() {
            val domainInstruction = metadata.stream().filter { instruction: Instruction? -> instruction is DomainInstruction }.findAny().orElse(null) as DomainInstruction?
            return domainInstruction!!.domain.split(" ").joinToString(""){ it.toString().capitalize() }
            return domainInstruction?.domain
        }

    var title: String? = null
        get() {
            val titleInstruction = metadata.stream().filter { instruction: Instruction? -> instruction is TitleInstruction }.findAny().orElse(null) as TitleInstruction?
            return titleInstruction?.title
        }

    var titleAsCapitalizedConcatainedString: String? = null
        get() {
            val titleInstruction = metadata.stream().filter { instruction: Instruction? -> instruction is TitleInstruction }.findAny().orElse(null) as TitleInstruction?
            return titleInstruction!!.title.split(" ").joinToString(""){ it.toString().capitalize() }
        }

    var version: String? = null
        get() {
            val versionInstruction = metadata.stream().filter { instruction: Instruction? -> instruction is VersionInstruction }.findAny().orElse(null) as VersionInstruction?
            return versionInstruction?.version
        }

    fun add(instruction: Instruction?) {
        metadata.add(instruction)
    }

    /*override fun getMetadata(): List<Instruction?> {
        return metadata
    }*/

    /*fun getTitle(): String? {
        val titleInstruction = metadata.stream().filter { instruction: Instruction? -> instruction is TitleInstruction }.findAny().orElse(null) as TitleInstruction?
        return titleInstruction?.title
    }*/

    /*override fun toString(): String {
        return "OrchaMetadata{" +
                "title='" + title + '\'' +
                ", description='" + description + '\'' +
                ", domain='" + domain + '\'' +
                ", author='" + author + '\'' +
                ", version='" + version + '\'' +
                '}'
    }*/
}
